File: NestCash/backend/app/models/message_models.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field
from beanie import Document, PydanticObjectId
from bson import ObjectId

class MessageDocument(Document):
    sender_id: PydanticObjectId
    receiver_id: PydanticObjectId
    sender_username: str  # Denormalizált adat
    receiver_username: str  # Denormalizált adat
    content: str
    is_read: bool = Field(default=False)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Settings:
        name = "private_messages"

class ConversationDocument(Document):
    participant_ids: List[PydanticObjectId] = Field(max_items=2, min_items=2)  # Pontosan 2 résztvevő
    participant_usernames: List[str] = Field(max_items=2, min_items=2)
    last_message_id: Optional[PydanticObjectId] = None
    last_message_content: Optional[str] = None
    last_message_at: Optional[datetime] = None
    unread_count_user1: int = Field(default=0)  # participant_ids[0] olvasatlan száma
    unread_count_user2: int = Field(default=0)  # participant_ids[1] olvasatlan száma
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Settings:
        name = "conversations"

# Response schemas
class MessageCreate(BaseModel):
    content: str = Field(..., min_length=1, max_length=1000)

class MessageRead(BaseModel):
    id: str
    sender_id: str
    receiver_id: str
    sender_username: str
    receiver_username: str
    content: str
    is_read: bool
    created_at: datetime
    is_my_message: bool = Field(default=False)

class ConversationRead(BaseModel):
    id: str
    other_user_id: str
    other_username: str
    last_message_content: Optional[str] = None
    last_message_at: Optional[datetime] = None
    unread_count: int = Field(default=0)

class ConversationListResponse(BaseModel):
    conversations: List[ConversationRead]
    total_count: int

class MessageListResponse(BaseModel):
    messages: List[MessageRead]
    total_count: int

File: NestCash/frontend/lib/models/message_models.dart
class PrivateMessage {
  final String id;
  final String senderId;
  final String receiverId;
  final String senderUsername;
  final String receiverUsername;
  final String content;
  final bool isRead;
  final DateTime createdAt;
  final bool isMyMessage;

  PrivateMessage({
    required this.id,
    required this.senderId,
    required this.receiverId,
    required this.senderUsername,
    required this.receiverUsername,
    required this.content,
    required this.isRead,
    required this.createdAt,
    required this.isMyMessage,
  });

  factory PrivateMessage.fromJson(Map<String, dynamic> json) {
    return PrivateMessage(
      id: json['id'],
      senderId: json['sender_id'],
      receiverId: json['receiver_id'],
      senderUsername: json['sender_username'],
      receiverUsername: json['receiver_username'],
      content: json['content'],
      isRead: json['is_read'] ?? false,
      createdAt: DateTime.parse(json['created_at']),
      isMyMessage: json['is_my_message'] ?? false,
    );
  }
}

class Conversation {
  final String id;
  final String otherUserId;
  final String otherUsername;
  final String? lastMessageContent;
  final DateTime? lastMessageAt;
  final int unreadCount;

  Conversation({
    required this.id,
    required this.otherUserId,
    required this.otherUsername,
    this.lastMessageContent,
    this.lastMessageAt,
    required this.unreadCount,
  });

  factory Conversation.fromJson(Map<String, dynamic> json) {
    return Conversation(
      id: json['id'],
      otherUserId: json['other_user_id'],
      otherUsername: json['other_username'],
      lastMessageContent: json['last_message_content'],
      lastMessageAt: json['last_message_at'] != null 
        ? DateTime.parse(json['last_message_at']) 
        : null,
      unreadCount: json['unread_count'] ?? 0,
    );
  }
}